### PR TITLE
[WIP] WorkerLoader V1.0.0-preview, OBJLoader2 V3.0.0-preview and wrappers

### PR DESCRIPTION
**Pull request by kaisalmen**
This was long in the making (#13663 and #13664) and it is not fully done yet, but I want to propose to introduce the `WorkerLoader` as a preview, so people can start testing it more easily and of course give feedback to the code.

I suggest to keep `OBJLoader2` V2.5.0 and `LoaderSupport` for now. The namespaces of the new and existing things are separated, so they can co-exist. `OBJLoader2` 3.0.0-preview is called `WLOBJLoader2` this shall reflect that it is used in conjunction with the new `WorkerLoader` and does no longer rely on `LoaderSupport` which means loaders are not required to extend it. Making a loader to work with `WorkerLoader` requires it to have a `parse` function (non-worker usage) and in addition a `buildWorkerCode` function for worker mode. Missing things can be added in a wrapper, see `WLPCDLoader`. `WorkerLoader` opens the door for auto-configured usage of loaders independent of worker usage.

The two base examples for `OBJLoader2` remain. The more complex ones and the new once for `DRACOLoader` (#15249) and `PCDLoader` rely on `WorkerLoader`.

This description is far from complete. I will add all the details in update to this post or a later post. Then it will also make sense to ping people.
....
**UPDATE:** I will not be able to write this down this before the end of the week.

I am currently moving to a new house that's why response time is likely higher and code commit activity is temporarily reduced, but I want to move this forward, finally.

### CODE REVIEW

File: examples/js/loaders/OBJLoader.js [hunk 1]
```diff
@@ -688,18 +688,14 @@ THREE.OBJLoader = ( function () {
 						if ( isLine && material && ! ( material instanceof THREE.LineBasicMaterial ) ) {
 
 							var materialLine = new THREE.LineBasicMaterial();
-							THREE.Material.prototype.copy.call( materialLine, material );
```
Comment (mrdoob) [inline examples/js/loaders/OBJLoader.js]: Did you modify this file by mistake?

Comment (kaisalmen) [reply examples/js/loaders/OBJLoader.js]: When I started my idea was to replace OBJLoader. I reverted in a later commit. Shall I squash parts of the history, so nobody gets confused?

Comment (Mugen87) [reply examples/js/loaders/OBJLoader.js]: Please revert these changes. Besides, I think it's no good idea to replace `OBJLoader`. From my point of view, `OBJLoader2` is not suitable to replace `OBJLoader` for different reasons (e.g. the general code style and architecture of the loader).

File: examples/files.js [hunk 2]
```diff
@@ -132,6 +130,12 @@ var files = {
 		"webgl_loader_vrm",
 		"webgl_loader_vrml",
 		"webgl_loader_vtk",
+		"webgl_loader_worker_gltf",
+		"webgl_loader_worker_pcd",
```
Comment (Mugen87) [inline examples/files.js]: I do not vote to add a PCD example. A single `glTF` example is sufficient from my point of view. We should keep in mind that we all have to maintain this code. Please keep the changes lean.

File: examples/js/loaders/OBJLoader.js (hunk 1 above)
Comment (Mugen87) [reply examples/js/loaders/OBJLoader.js]: Sorry for the lack of feedback but the `OBJ` format is not that important compared to other 3D formats these days.

In any event, `OBJLoader2` does things like input validation (the usage of `THREE.LoaderSupport.Validator`). In general, `three.js` does not perform something like that for performance reasons. I also think that the general handling of `LoaderSupport` support feels like over-engineering to me. You clearly overestimate the benefit of workers in context of loading 3D assets. From my point of view we don't need a generic worker solution for loaders. The only valid use case for me is in context of `Draco` decompression right now.

**Comment by kaisalmen** [comment]
Hello, I finally woke up from my winter sleep. Sorry for staying silent for much longer than I personally expected. I really started to work on this whole subject again in my [WWOBJLoader dev branch](https://github.com/kaisalmen/WWOBJLoader/tree/dev) repo.

Updated goals are:
- [x] Make OBJLoader2 completely self-standing (no more Validator code deps, asset handling internally), `OBJLoader2.Parser` has no dependencies to `THREE` so it is able to run in Worker with least possible foot-print
- [ ] `WorkerLoader` is actually an Batch/Automation-Loader with possibility to wrap Loaders in Workers or drive existing loaders from external configuration (with code or json/yaml in the future). This should be reflected by name and composition of the different pieces (have smaller less code dependent blocks).
- [ ] Chain different loaders and make assets available from one to the next in a generic way. This could remove `MTLLoader` wrapper code from `OBJLoader2`.
- [ ] Make assets available from Main to Worker bi-directional = evolve `MeshTransfer` to `AssetTransfer`
- [ ] Modernize code / Transform to jsm

As soon as I have something working again, I will rebase this branch and add new commits.

Are you ok with this PR staying open or should I close it and start a new one later.

Thank you.

**Comment by Mugen87** [comment]
> Are you ok with this PR staying open or should I close it and start a new one later.

I think it's better to close it and start with a fresh one 👍

**Pull request closed without merging**